### PR TITLE
feat: update key for client

### PR DESCRIPTION
**Pull request by yanganto**
Hi there,
This PR is a reopen one for reloading key
https://github.com/near/nearcore/pull/8117

When the SIGHUP signal receives and `update_key_file: true` is set.  We want to insert a key and turn the node into a validator.

If there is something still unsuitable in this PR, please kindly point it out and let me know.  Thank you. :pray:

### CODE REVIEW

**Comment by nikurt** [comment]
Nice to see you back with a new PR on the topic of switching the validator key.
I checked out your branch. Let me play with it, and I'll get back to you with feedback in a couple of days.

This may need an integration test.

And I'm unsure about the field `update_key_file` in the config. It seems to be either `None` or `true`. Let me think about it.

**Comment by yanganto** [comment]
@nikurt Thank you for taking care of this PR. This is really subtle, and really need experience with Near network, so it is hard to say whether there is a bug somewhere or not.
Meanwhile, this PR can turn a backup node into a validator in a really quick way, so I think this is useful.  If you have any concerns, it is happy to have advice from your end.

**Comment by yanganto** [comment]
By the way, there is also a corresponding PR on Kuutamo with some integration tests on it.
https://github.com/kuutamolabs/near-staking-knd/pull/614

**Comment by nikurt** [comment]
Hm, tried it on localnet, and the node failed to assume the identity of a validator. Need more time to investigate.

**Comment by yanganto** [comment]
Oh, I think the `node_key` should be updated at the same time, such that this issue can be solved.

**Comment by yanganto** [comment]
Thanks, I am checking on this.

File: nearcore/src/dyn_config.rs [hunk 1]
```diff
@@ -45,7 +46,14 @@ pub fn read_updateable_configs(
 pub fn get_updateable_client_config(config: Config) -> UpdateableClientConfig {
     // All fields that can be updated while the node is running should be explicitly set here.
     // Keep this list in-sync with `core/dyn-configs/README.md`.
-    UpdateableClientConfig { expected_shutdown: config.expected_shutdown }
+    UpdateableClientConfig {
+        expected_shutdown: config.expected_shutdown,
+        validator_signer: if config.update_key_file.unwrap_or_default() {
+            InMemoryValidatorSigner::from_file(&Path::new(&config.validator_key_file)).ok()
```
Comment (nikurt) [inline nearcore/src/dyn_config.rs:52]: This is looking for the validator key file in the wrong directory.
Should be `home_dir.join(config.validator_key_file)`.

**Comment by nikurt** [comment]
Also `Client::block_production_started` needs to be set to `true` when `Client::validator_signer` is set.

File: nearcore/src/dyn_config.rs (hunk 1 above)
Comment (yanganto) [reply nearcore/src/dyn_config.rs]: Agree, the absolute path should be used here.
Thanks for pointing that out.

**Comment by yanganto** [comment]
@nikurt Thank you. Still need some work on the PR, so it turn back into a draft now.

**Comment by nikurt** [comment]
cc: @VanBarbascu

**Comment by paulmattei** [comment]
Hey folks, are we able to progress this?

**Review by VanBarbascu — COMMENTED** [review]
Hi Antonio!
Can you provide some test steps?

I tried to use this feature on my testnet with 4 validator nodes. (nodes 0 to 3)

I stopped node3, removed the `validator_key.json`, added the `update_key_file` field to `config.json` and then restarted node3.
Current validators are node0-2 and node3 is just following the network.
To hand over node2 to node3, I copied `validator_key.json` from 2 to 3. killed node2 and sighup-ed node3 to reload the key.

The node was able to load the key but it was still missing to produce blocks.

File: nearcore/src/dyn_config.rs [hunk 2]
```diff
@@ -42,10 +43,17 @@ pub fn read_updateable_configs(
     }
 }
 
-pub fn get_updateable_client_config(config: Config) -> UpdateableClientConfig {
+pub fn get_updateable_client_config(home_dir: &Path, config: Config) -> UpdateableClientConfig {
     // All fields that can be updated while the node is running should be explicitly set here.
     // Keep this list in-sync with `core/dyn-configs/README.md`.
```
[the remark this answers is not in the document]
Comment (yanganto) [reply nearcore/src/dyn_config.rs:48]: Got it, thanks for review. :)

**Comment by VanBarbascu** [comment]
Hi @yanganto, 
I am checking back on this PR to see if you have any update on the testing flow.

**Comment by VanBarbascu** [comment]
When you come back to this, please take a look at the discussion we had in [this thread](https://near.zulipchat.com/#narrow/stream/297663-pagoda.2Fnetwork/topic/Changing.20IP.20of.20validator/near/364425985) about what it means form network perspective to change the ip and node identity.

**Comment by yanganto** [comment]
Got it. Thank you.  :pray: :+1:

**Comment by akhi3030** [comment]
looks like this PR is not being worked on actively.  Closing till the authors are ready to come back to it.

**Pull request closed without merging**